### ISSUE TITLE
CH16: nan loss with stateful RNN

### ISSUE THREAD

**Issue by thierryherrmann** [ageron/handson-ml2#32]
I'm unable to reproduce the output of section `Stateful RNN` in https://github.com/ageron/handson-ml2/blob/master/16_nlp_with_rnns_and_attention.ipynb

I've executed all cells of the notebook but when executing this cell:
```
model.compile(loss="sparse_categorical_crossentropy", optimizer="adam")
steps_per_epoch = train_size // batch_size // n_steps
model.fit(dataset, steps_per_epoch=steps_per_epoch, epochs=50,
                   callbacks=[ResetStatesCallback()])
```
I quickly end up with a nan loss:
```
Train for 313 steps
Epoch 1/50
 38/313 [==>...........................] - ETA: 45s - loss: nan
```
Instead of the output committed in the notebook:
```
Epoch 1/50
313/313 [==============================] - 101s 322ms/step - loss: 2.6180
Epoch 2/50
313/313 [==============================] - 98s 312ms/step - loss: 2.2312
...etc...
```
I'm using:
```
tensorflow  2.0.0rc0
numpy       1.17.0
```
on ubuntu 18.04.

I've tried to find the problem by using a manual training loop (with `GradientTape`...) and batches of 1 instances. I could reproduce the exact same results I get with `model.fit()`, but I can also access intermediate layers, check the GRU weights, dense layer weights and outputs and the gradients. After learning with a few examples the 2nd GRU layer (upper) outputs large numbers (and gradually increasing as the time step increases), leading to large outputs in the dense layer, nan in the softmax output, hence nan loss. But I'm unable to explain why the network evolves in that direction.
I've added regularizers to the GRUs but they only slightly delay the problem.
I haven't tried with tensorflow `2.0.0-beta1`.

Am I the only one ? When this kind of things happens, what's the best path to understand the problem (and maybe solve it) ?

For the other RNN (non-stateful) (section `Creating and Training the Model`) I obtained the same result as the one committed in the notebook.

**Comment by ageron** [comment]
Hi @thierryherrmann ,

Thanks for your feedback. I’m sorry you ran into this issue. The most common cause of NaNs during training is an excessively large learning rate. Could you please try reducing it (eg by a factor of 10) to see if that's the cause of the problem. RNNs can be unstable, so if anything is different (e.g., the random number generator's seed) then things can go very differently. With RNNs, it's common to have to start the first few epochs with a very small learning rate, and then increase it a bit.
I'll try to run the notebook and I'll check whether I get the same problem now (perhaps something changed in TensorFlow since I ran the code).

**Comment by ageron** [comment]
Mmh, I was able to reproduce this issue on Colab. I tried reducing the learning rate by a factor of 10 and even 1000, but it bombs anyway. Something's fishy... I'm investigating !

**Comment by ageron** [comment]
Just confirmed the issue on my MacBook using the latest libraries. I'll try to revert to TensorFlow from May 10th (when the code ran successfully), and I'll check whether it works.

**Comment by ageron** [comment]
Wow, I tried running the notebook again using a nightly build of TensorFlow from May 10th, and it behaves the same way (i.e., NaN after a short time). So it seems that the issue is not due to a TensorFlow update. I really don't understand how I managed to run this. Perhaps I ran it on a GPU, which makes the outputs non-deterministic, and I was lucky enough to get a good run without NaNs? I'll keep looking.

**Comment by thierryherrmann** [comment]
Note that I've tried without dropout:

```python
model = keras.models.Sequential([
    keras.layers.GRU(128, return_sequences=True, stateful=True,
                     #dropout=0.2, recurrent_dropout=0.2,
                     batch_input_shape=[batch_size, None, max_id]),
    keras.layers.GRU(128, return_sequences=True, stateful=True,
                     #dropout=0.2, recurrent_dropout=0.2
                    ),
    keras.layers.TimeDistributed(keras.layers.Dense(max_id,
                                                    activation="softmax"))
])
```
and the outputs of the upper GRU layer no longer explode as time steps increase and there's no longer nan loss. The training went successfully through the 313 epochs.  (Hmmm....)
```
Train for 313 steps
Epoch 1/50
313/313 [==============================] - 26s 83ms/step - loss: 2.4841
Epoch 2/50
313/313 [==============================] - 26s 82ms/step - loss: 1.9857
Epoch 3/50
313/313 [==============================] - 25s 81ms/step - loss: 1.7911
Epoch 4/50
313/313 [==============================] - 24s 77ms/step - loss: 1.6789

...

Epoch 48/50
313/313 [==============================] - 24s 77ms/step - loss: 1.2179
Epoch 49/50
313/313 [==============================] - 24s 78ms/step - loss: 1.2159
Epoch 50/50
313/313 [==============================] - 24s 78ms/step - loss: 1.2141

```

**Comment by ageron** [comment]
Ooooh that's a fantastic insight, thanks! You just made me remember that I filed [an issue about `recurrent_dropout` not working in stateful models](https://github.com/tensorflow/tensorflow/issues/27829) when I was working on this chapter in May. Since dropout didn't work, perhaps I ran training without it, and later when the issue was fixed (or at least closed), I added dropout back in the code but I never got around to actually retrain the model. If that's what happened, it was a bit clumsy of me, and I apologize for the inconvenience.

Now we need to figure out whether dropout is broken in stateful model in TensorFlow, or whether it's just making the model unstable in this particular case. Unfortunately I'll be traveling over the next 2 weeks, and I won't be able to work on this until I get back.

Hopefully, this sheds a bit more light on the problem.

**Comment by thierryherrmann** [comment]
Great! Thanks for the explanation.

> never got around to actually retraining the model

This is what I suspected. Until [the issue](https://github.com/tensorflow/tensorflow/issues/27829) is fixed, you might want to remove the dropout from the notebook, at least temporarily. We're not sure dropout helps anyway. I can do a PR if you want. Up to you.
Big thanks again for your extraordinary book and great notebooks.

**Comment by ageron** [comment]
Thanks @thierryherrmann , I agree! A PR would be fantastic: please just comment it out and maybe point to this issue in the comment, since the book unfortunately contains the dropout line. Thanks again! 👍

**Comment by thierryherrmann** [comment]
Done! Thanks! 
https://github.com/ageron/handson-ml2/pull/34

**Comment by ageron** [comment]
Closing since the root issue is closed. Thanks again for your help @thierryherrmann.
